---
title: "WGCNA_Sid_Host"
author: "Veronica Radice"
date: "22/03/2021"
output: html_document
---

# Coral Host *Siderastrea*

# WGCNA
## Weighted Gene Co-Expression Network Analysis

Correlation networks are increasingly being used in bioinformatics applications. For example, weighted gene co-expression network analysis is a systems biology method for describing the correlation patterns among genes across microarray samples. Weighted correlation network analysis (WGCNA) can be used for finding clusters (modules) of highly correlated genes, for summarizing such clusters using the module eigengene or an intramodular hub gene, for relating modules to one another and to external sample traits (using eigengene network methodology), and for calculating module membership measures. Correlation networks facilitate network based gene screening methods that can be used to identify candidate biomarkers or therapeutic targets. These methods have been successfully applied in various biological contexts, e.g. cancer, mouse genetics, yeast genetics, and analysis of brain imaging data. While parts of the correlation network methodology have been described in separate publications, there is a need to provide a user-friendly, comprehensive, and consistent software implementation and an accompanying tutorial.

The WGCNA R software package is a comprehensive collection of R functions for performing various aspects of weighted correlation network analysis. The package includes functions for network construction, module detection, gene selection, calculations of topological properties, data simulation, visualization, and interfacing with external software.

https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/index.html

https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html

### References

Package, and publications:

  Langfelder P and Horvath S, WGCNA: an R package for weighted correlation network
  analysis. BMC Bioinformatics 2008, 9:559 doi:10.1186/1471-2105-9-559

  Peter Langfelder, Steve Horvath (2012). Fast R Functions for Robust Correlations and
  Hierarchical Clustering. Journal of Statistical Software, 46(11), 1-17. URL
  http://www.jstatsoft.org/v46/i11/.

  Bin Zhang and Steve Horvath (2005) "A General Framework for Weighted Gene Co-Expression Network Analysis", Statistical Applications in Genetics and Molecular Biology: Vol. 4: No. 1, Article 17
 
  Horvath S, Dong J (2008) Geometric Interpretation of Gene Coexpression Network Analysis. PLoS Comput Biol 4(8): e1000117
 
 
## Import libraries
```{r message= FALSE}
## If don't have BiocManager library installed:
# install.packages("BiocManager")

## For first time installation:
# library(BiocManager)
# BiocManager::install("WGCNA")
# BiocManager::install("impute")
# BiocManager::install("GO.db")
# BiocManager::install("preprocessCore")

library(DESeq2)
# or
#library(SummarizedExperiment) for assay()

library(flashClust)
library(WGCNA)
library(impute)
library(GO.db)
library(preprocessCore)
library(dplyr)
library(plyr)
library(tidyverse)

options(stringsAsFactors=FALSE)
allowWGCNAThreads()  # Allowing multi-threading with up to 8 threads. nThreads =
enableWGCNAThreads()  # Allowing parallel execution with up to 7 working processes.
```


## Set working directory
```{r}
setwd("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/Siderastrea/host/")
```


## Import metadata
```{r}
meta <- read.csv("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/Siderastrea/Siderastrea_Year-2_metadata.csv", header=TRUE, row.names=1)

# Remove all samples that were transplanted to the Destination Ojo.Norte
# There is no control site for Norte Ojo
# For now, we are focused on samples transplanted to Laja Control or Laja Ojo sites
# This is consistent with analyses for the other 2 species in the experiment
# Also, group sample size is too small for Ojo.Norte destination

meta <- subset(meta, Destination_name != "Ojo.Norte")
```


```{r}
meta$Destination_type <- revalue(meta$Destination_type, c("control" = "Control", "ojo" = "Ojo"))

meta$Origin_type <- revalue(meta$Origin_type, c("Control" = "Lagoon"))

meta <- meta %>% unite(group, Origin_type, Destination_type, sep = ".", remove = FALSE)


meta[sapply(meta, is.character)] <- lapply(meta[sapply(meta, is.character)], 
                                       as.factor)
# Colony_ID is equivalent to genotype
meta$Colony_ID <- as.factor(meta$Colony_ID)

levels(meta$group)
```


# select relevant metadata
```{r}
meta <- dplyr::select(meta, Colony_ID, group, Origin_name, Origin_type, Destination_name, Destination_type, pH_Destination)
```

### Remove samples with >85% of null/zero counts
As in DESeq2 analysis
```{r}
drop <- c("R_026_N_Sid_yr2") # this is an Ojo Norte sample so already removed above

meta <- meta[!(row.names(meta) %in% drop),]
meta <- droplevels(meta)
```

### Remove outlier
As in DESeq2 analysis
```{r}
drop <- c("R_018_C_Sid_yr2")

meta <- meta[!(row.names(meta) %in% drop),]
meta <- droplevels(meta)
dim(meta)
```

### data summary
```{r}
meta %>%
  group_by(group) %>%
  dplyr::summarise(count = n())
```

####################################################################

# Import transformed counts

The input file has to contain ALL the genes, not just differentially expressed ones.
```{r}
counts.transformed.vst <- readRDS(file = "counts_vst.BlindFalse_Sid_Host.rds")
```


```{r}
dat = as.data.frame(assay(counts.transformed.vst))
head(dat)
```

### Transpose counts
```{r}
dat_t <- data.table::transpose(dat)
rownames(dat_t) <- colnames(dat)
colnames(dat_t) <- rownames(dat)
```


## Import trait data
```{r}
traits <- read.csv("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/Siderastrea/Martinez_2019_Caribbean-Coral-transplant-Calcification-physiol-data.csv", header=TRUE, row.names=1)

traits <- subset(traits, species == "Siderastrea")

traits$Destination_type <- revalue(traits$Destination_type, c("control" = "Control", "ojo" = "Ojo"))
traits$Destination_type <- factor(traits$Destination_type, levels = c("Control", "Ojo"))

traits$origin <- revalue(traits$origin, c("Control" = "Lagoon"))
traits$origin <- factor(traits$origin, levels = c("Lagoon", "Reef", "Ojo"))

traits <- traits %>% unite(group, origin, Destination_type, sep = ".", remove = FALSE)

traits <- traits %>% unite(treatment_sites, Origin_name, Destination_name, sep = "_", remove = FALSE)

traits <- traits %>% unite(name, Origin_name, Destination_name, Colony_ID, sep = "_", remove = FALSE)

traits <- dplyr::rename(traits, treatment = treatment_Origin.Dest)

traits[sapply(traits, is.character)] <- lapply(traits[sapply(traits, is.character)], 
                                       as.factor)
# Colony_ID is equivalent to genotype
traits$Colony_ID <- as.factor(traits$Colony_ID)

traits <- traits %>% unite(colonyID_sampleID, Origin_type, Destination_type, Colony_ID, id, sep = ".", remove = FALSE)

levels(traits$group)
```

### Remove outlier
As in DESeq2 analysis
```{r}
drop <- c("R_018_C_Sid_yr2")

traits <- traits[!(row.names(traits) %in% drop),]
traits <- droplevels(traits)
dim(traits)
```

# for checking dendrogram sample clustering by colony (genotype)
```{r}
dat2 <- data.table::transpose(dat, keep.names=NULL)
colnames(dat2) <- rownames(dat)
rownames(dat2) <- traits$colonyID_sampleID
head(rownames(dat2))
```


# select relevant trait data
```{r}
traits <- dplyr::select(traits, zoox.cells.cm2, chla.ug.cm2, chla.ug.cell, protein.mg.cm2, colonyID_sampleID)

# 10^6 measurements are same data just different units as needed for plotting
# traits <- dplyr::select(traits, zoox.cells.cm2, zoox.10.6.cells.cm2,	chla.ug.cm2,	chla.10.6.ug.cm2, chla.ug.cell,	protein.mg.cm2)
```


WGCNA starts from the level of thousands of genes, identifies clinically interesting gene modules, and finally uses intramodular connectivity, gene significance (e.g. based on the correlation of a gene expression profile with a sample trait) to identify key genes in the disease pathways for further validation. 

WGCNA alleviates the multiple testing problem inherent in microarray data analysis. Instead of relating thousands of genes to a microarray sample trait, it focuses on the relationship between a few (typically less than 10) modules and the sample trait. Toward this end, it calculates the eigengene significance (correlation between sample trait and eigengene) and the corresponding p-value for each module. The module definition does not make use of a priori defined gene sets. Instead, modules are constructed from the expression data by using hierarchical clustering. 

Modules are clusters of highly interconnected genes.

Although it is advisable to relate the resulting modules to gene ontology information to assess their biological plausibility, it is not required. Because the modules may correspond to biological pathways, focusing the analysis on intramodular hub genes (or the module eigengenes) amounts to a biologically motivated data reduction scheme. Because the expression profiles of intramodular hub genes are highly correlated, typically dozens of candidate biomarkers result. Although these candidates are statistically equivalent, they may differ in terms of biological plausibility or clinical utility. Gene ontology information can be useful for further prioritizing intramodular hub genes.


**WGCNA requires genes as columns**
```{r}
# get row and colnames in order
colnames(dat_t) <- rownames(dat)
rownames(dat_t) <- colnames(dat)
dat <- dat_t
dim(dat)
```

### Iterative filtering of samples and genes with too many missing entries

This function checks data for missing entries, entries with weights below a threshold, and zero-variance genes, and returns a list of samples and genes that pass criteria on maximum number of missing or low weight values. If necessary, the filtering is iterated.
```{r}
gsg = goodSamplesGenes(dat, verbose = 3);
gsg$allOK
#if TRUE, no outlier genes
```

## Check data for excessive missing values and identification of outliers
```{r}
# good sample genes
if (!gsg$allOK)
{
  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg$goodGenes)>0) 
     printFlush(paste("Removing genes:", paste(names(dat)[!gsg$goodGenes], collapse = ", ")));
  if (sum(!gsg$goodSamples)>0) 
     printFlush(paste("Removing samples:", paste(rownames(dat)[!gsg$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  dat = dat[gsg$goodSamples, gsg$goodGenes]
}
dim(dat)
```

## Visualize outliers - by Sample ID
```{r}
sampleTree = hclust(dist(dat), method = "average");
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
sizeGrWindow(12,9)
#pdf(file = "Plots/sampleClustering.pdf", width = 12, height = 9);
par(cex = 0.6);
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Siderastrea Host - Sample ID", sub="", xlab="", cex.lab = 1.5, 
     cex.axis = 1.5, cex.main = 2)
```


## Visualize outliers - by Colony ID
```{r}
sampleTree = hclust(dist(dat2), method = "average");
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
sizeGrWindow(12,9)
#pdf(file = "Plots/sampleClustering.pdf", width = 12, height = 9);
par(cex = 0.6);
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Siderastrea Host - Colony ID", sub="", xlab="", cex.lab = 1.5, 
     cex.axis = 1.5, cex.main = 2)
```

## Remove outlier
```{r}
drop <- c("R_024_C_Sid_yr2")

meta <- meta[!(row.names(meta) %in% drop),]
traits <- traits[!(row.names(traits) %in% drop),]
dat <- dat[!(row.names(dat) %in% drop),]

# drop factor levels
traits[] <- lapply(traits, function(x) if(is.factor(x)) factor(x) else x)

meta <- droplevels(meta)
traits <- droplevels(traits)
dat <- droplevels(dat)

dim(traits)
dim(dat)

dat3 <- dat2[-33,]
dim(dat3)
```

```{r}
table(rownames(traits)==rownames(dat)) 
# should return TRUE if datasets align correctly, otherwise your names are out of order
```


## Dendrogram - by sample ID
```{r}
sampleTree = hclust(dist(dat), method = "average")
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
sizeGrWindow(12,9)
par(cex = 0.6)
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Siderastrea Host",  sub="", xlab="", cex.lab = 1.5, 
     cex.axis = 1.5, cex.main = 2) 
```

save file
```{r}
pdf(file = "Sid_Host_dendrogram_SampleID.pdf", width = 13, height = 9)
par(cex = 0.6)
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Siderastrea Host",  sub="", xlab="", cex.lab = 1.5, 
     cex.axis = 1.5, cex.main = 2) 

dev.off()
```


## Dendrogram - by Colony ID
```{r}
sampleTree = hclust(dist(dat3), method = "average")
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
sizeGrWindow(12,9)

# save file
#pdf(file = "Sid_Host_dendrogram_ColonyID_SampleID.pdf", width = 12, height = 9)
par(cex = 0.6)
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Siderastrea Host", sub="", xlab="", cex.lab = 1.5, 
     cex.axis = 1.5, cex.main = 2)

#dev.off()
```


## Remove non-relevant, non-numeric columns left in traits dataframe
```{r}
traits <- dplyr::select(traits, zoox.cells.cm2, chla.ug.cm2, chla.ug.cell, protein.mg.cm2)

# 10^6 measurements are same data just different units as needed for plotting
# traits <- dplyr::select(traits, zoox.cells.cm2, zoox.10.6.cells.cm2,	chla.ug.cm2,	chla.10.6.ug.cm2, chla.ug.cell,	protein.mg.cm2)
```



```{r}
table(rownames(traits)==rownames(dat)) 
# should return TRUE if datasets align correctly, otherwise your names are out of order

# You have finished uploading and formatting expression and trait data
save(dat, traits, file="Sid_Host_SamplesAndTraits.RData")
#load("Sid_Host_SamplesAndTraits.RData")
```


## Visualize how the traits relate to the sample dendrogram
```{r message=FALSE}
# cluster samples
sampleTree2 = hclust(dist(dat), method = "average")
# as above, dat is transposed rlogged counts

# Convert traits to a color representation: white means low, red means high, grey means missing entry
traitColors = numbers2colors(traits, signed = FALSE)

# Plot the sample dendrogram and the colors underneath.
plot <- plotDendroAndColors(sampleTree2, traitColors,
                    groupLabels = names(traits), 
                    main = "Dendrogram with Trait heatmap")

plot
```

save plot
```{r}
pdf("Sid_Host_dendrogram_Trait-heatmap.pdf") 

# cluster samples
sampleTree2 = hclust(dist(dat), method = "average")
# as above, dat is transposed rlogged counts

# Convert traits to a color representation: white means low, red means high, grey means missing entry
traitColors = numbers2colors(traits, signed = FALSE)

# Plot the sample dendrogram and the colors underneath.
plot <- plotDendroAndColors(sampleTree2, traitColors,
                    groupLabels = names(traits), 
                    main = "Dendrogram with Trait heatmap")

plot

dev.off() 
```



####################################################################

# 1. Construct a gene co-expression network
Rationale: make use of interaction patterns among genes 
Tools: correlation as a measure of co-expression

Constructing a weighted gene network entails the choice of the soft thresholding power B (beta) to which co-expression similarity is raised to calculate adjacency.


## Network construction and module detection

Choosing a soft-threshold to fit a scale-free topology to the network
```{r}
powers = c(c(1:10), seq(from = 12, to=20, by=2))
```


### Signed or unsigned: which network type is preferable?

https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html

The choice of signed vs. unsigned networks is complex, but in general we prefer signed (or "signed hybrid") networks to unsigned networks. 

https://peterlangfelder.com/2018/11/25/signed-or-unsigned-which-network-type-is-preferable/

How should pairs of nodes with strong negative correlations be treated in a correlation network analysis? 

- One option is to consider them connected, just as if the correlation were positive. A network constructed in this way is an unsigned network, because the sign of the correlation does not matter. 
- On the other hand, strongly negatively correlated nodes can also be considered unconnected. This leads to a signed network, so called because the sign of a strong correlation value makes all the difference between the pair of nodes being strongly connected or not connected at all. To avoid any confusion, I want to emphasize that the resulting adjacency matrix (the matrix that contains the connection strengths between nodes) is always non-negative.

Should you use a signed or unsigned network? 

*By and large, I recommend using one of the signed varieties, for two main reasons.* 
1.  First, more often than not, direction does matter: it is important to know where node profiles go up and where they go down, and mixing negatively correlated nodes together necessarily mixes the two directions together. 

2.  Second, negatively correlated nodes often belong to different categories. For example, in gene expression data, negatively correlated genes tend to come from biologically very different categories. It is true that some pathways or processes involve pairs of genes that are negatively correlated; if there are enough negatively correlated genes, they will form a module on their own and the two modules can then be analyzed together. 

(For the advanced practitioner, another option is to use the fuzzy module membership measure based on the module eigengene to attach a few strongly negatively correlated genes to a module after the modules have been identified).


#### A variance-stabilizing transformation is recommended
For example, package DESeq2 implements the function varianceStabilizingTransformation which we have found useful, but one could also start with normalized counts (or RPKM/FPKM data) and log-transform them using log2(x+1). For highly expressed features, the differences between full variance stabilization and a simple log transformation are small. 


***! Apparently pickSoftThreshold() is known NOT TO WORK IN RSTUDIO !***
Moved to R for this specific chunk
```{r}
# library(SummarizedExperiment)
# library(DESeq2)
# library(WGCNA)
# options(stringsAsFactors=FALSE)
# enableWGCNAThreads()
# 
# setwd("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/Siderastrea/host/")
# counts.transformed.vst <- readRDS(file = "counts_vst.BlindFalse_Sid_Host.rds")
# 
# dat = as.data.frame(assay(counts.transformed.vst))
# dat_t <- data.table::transpose(dat)
# colnames(dat_t) <- rownames(dat)
# rownames(dat_t) <- colnames(dat)
# dat <- dat_t
# dim(dat)
# 
# drop <- c("R_024_C_Sid_yr2")
# dat <- dat[!(row.names(dat) %in% drop),]
# dat <- droplevels(dat)
# dim(dat)
# 
# drop <- c("R_018_C_Sid_yr2")
# dat <- dat[!(row.names(dat) %in% drop),]
# dat <- droplevels(dat)
# dim(dat)

## Choose a set of soft-thresholding powers
  # powers = c(seq(4,10,by=1), seq(12,20, by=2))

## Call the network topology analysis function
## The function calculates weighted networks either by interpreting data directly as similarity, or first transforming it to similarity of the type specified by networkType. The weighted networks are obtained by raising the similarity to the powers given in powerVector. For each power the scale free topology fit index is calculated and returned along with other information on connectivity.

  # sft = pickSoftThreshold(dat, dataIsExpr = TRUE, powerVector = powers, networkType = "signed", verbose = 2)
  # saveRDS(sft, file = "Sid_Host_SoftThreshold.rds")
```
pickSoftThreshold: will use block size 5592.
 pickSoftThreshold: calculating connectivity for given powers...
   ..working on genes 1 through 5592 of 8000
   ..working on genes 5593 through 8000 of 8000
   Power SFT.R.sq slope truncated.R.sq mean.k. median.k. max.k.
1      4   0.0697 -1.69          0.934  700.00    687.00 1080.0
2      5   0.3350 -2.99          0.943  420.00    406.00  777.0
3      6   0.5780 -3.67          0.960  260.00    246.00  578.0
4      7   0.7090 -3.79          0.971  165.00    152.00  441.0
5      8   0.7970 -3.69          0.985  108.00     96.20  345.0
6      9   0.8270 -3.56          0.988   72.00     62.00  274.0
7     10   0.8560 -3.40          0.992   49.10     40.60  221.0
8     12   0.8810 -3.13          0.991   24.20     18.20  150.0
9     14   0.8940 -2.89          0.993   12.80      8.62  105.0
10    16   0.8900 -2.65          0.980    7.21      4.27   76.1
11    18   0.8970 -2.34          0.965    4.29      2.20   56.5
12    20   0.9560 -1.99          0.987    2.69      1.17   42.8


```{r}
# read in SoftPower object generated in R
sft <- readRDS("Sid_Host_SoftThreshold.rds")

# Plot the results:
sizeGrWindow(9, 5)
par(mfrow = c(1,2))
cex1 = 0.9

# Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
     main = paste("Scale independence"));
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col="red")
# this line corresponds to using an R^2 cut-off of h
abline(h=0.90,col="red")

# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
     main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
```

**Looking for:**
1. the scale-free topology fit index to reach values above 0.8 for reasonable powers 
  - usually it is recomended to choose the lowest softPower that has an R2 at 90%
  - less than 15 for unsigned or signed hybrid networks, and less than 30 for signed networks 
2. the mean connectivity to remain relatively low 
  - high is considered in the hundreds or above


If the scale-free topology fit index fails to reach values above 0.8 for reasonable powers (less than 15 for unsigned or signed hybrid networks, and less than 30 for signed networks) and the mean connectivity remains relatively high (in the hundreds or above), chances are that the data exhibit a strong driver that makes a subset of the samples globally different from the rest. The difference causes high correlation among large groups of genes which invalidates the assumption of the scale-free topology approximation. 


### Calculate (correlation or distance) network adjacency from given expression data
```{r}
# Generate adjacency and TOM similarity matrices based on the selected softpower
softPower = 9
adjacency = adjacency(dat, type = "signed", power = softPower)
# corFnc defaults to Pearson correlation
```


### Topological overlap matrix (TOM)
```{r}
# turn adjacency matrix into topological overlap to minimize the effects of noise and spurious associations
TOM = TOMsimilarityFromExpr(dat, networkType = "signed", TOMType = "signed", power = softPower)

colnames(TOM) = rownames(TOM)
dissTOM = 1-TOM
```


####################################################################

# 2. Identify modules

Rationale: module (pathway) based analysis 
Tools: hierarchical clustering, Dynamic Tree Cut

# Module detection
```{r}
# hierarchical clustering of the genes based on the TOM dissimilarity measure
geneTree = flashClust(as.dist(dissTOM), method="average")

# plot the resulting clustering tree (dendrogram)
plot(geneTree, xlab="", sub="", cex=0.3, labels = FALSE, hang = 0.04)
```


# Adaptive Branch Pruning of Hierarchical Clustering Dendrograms

## Clustering using dynamic tree cut

Two methods for hierarchical clustering:

- method="tree"
Dynamic tree cut is a top-down algorithm that relies solely on the dendrogram. The algorithm implements an adaptive, iterative process of cluster decomposition and combination and stops when the number of clusters becomes stable.

- method="hybrid"
Dynamic hybrid cut is a bottom-up algorithm that improves the detection of outlying members of each cluster. This variant can be considered a hybrid of hierarchical clustering and modified partitioning around medoids (PAM).

*Here I use method = "hybrid"*


## Dynamic *hybrid* cut (bottom-up)

Module identification using dynamic tree cut:

- cutHeight - Maximum joining heights that will be considered.
  - setting cutHeight =.9 means the maximum dissimilarity is (90%) that qualifies modules for merging.
  - For method=="tree" it defaults to 0.99. 
  - For method=="hybrid" it defaults to 99% of the range between the 5th percentile and the maximum of the joining heights on the dendrogram.
- distM - Only used for method "hybrid"
- deepSplit 
  - For method "hybrid", can be either logical or integer in the range 0 to 4. For method "tree", must be logical. In both cases, provides a rough control over sensitivity to cluster splitting. The higher the value (or if TRUE), the more and smaller clusters will be produced. For the "hybrid" method, a finer control can be achieved via maxCoreScatter and minGap below.
- pamRespectsDendro
  - Logical, only used for method "hybrid". If TRUE, the PAM stage will respect the dendrogram in the sense that objects and small clusters will only be assigned to clusters that belong to the same branch that the objects or small clusters being assigned belong to.

```{r}
# Adjust the minimum module size, 30 is common as it is what is suggested in the official tutorials.
minModuleSize = 30

dynamicMods_hybrid = cutreeDynamic(dendro = geneTree, method="hybrid", distM = dissTOM,
                            deepSplit = 2, pamRespectsDendro = FALSE,
                            minClusterSize = minModuleSize, verbose = 3)
```

gives the module labels and the size of each module
```{r}
table(dynamicMods_hybrid)
```


Convert numeric labels into colors
```{r}
dynamicColors = labels2colors(dynamicMods_hybrid)
table(dynamicColors)
```

## Plot the module assignment under the dendrogram
```{r}
sizeGrWindow(8,6)
plotDendroAndColors(geneTree, dynamicColors, "Dynamic Hybrid Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene dendrogram and module colors")
```

### Calculate eigengenes
```{r}
MEList = moduleEigengenes(dat, colors = dynamicColors)
MEs = MEList$eigengenes

# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs)

# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = "average")

# Plot the result
sizeGrWindow(7, 6)
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")

MEDissThres = 0.15 # merge modules that are 85% similar
abline(h = MEDissThres, col="red")
```


```{r}
save(dynamicMods_hybrid, MEList, MEs, MEDiss, METree, file= "Sid_Host_WGCNA_ME_Network.RData")
```


### Merging of modules whose expression profiles are very similar
  
The Dynamic Tree Cut may identify modules whose expression profiles are very similar. 
It may be prudent to merge such modules since their genes are highly co-expressed. 
To quantify co-expression similarity of entire modules, we calculate their eigengenes (MEs) and cluster them on their consensus correlation.
```{r}
merge = mergeCloseModules(dat, dynamicColors, cutHeight = MEDissThres, verbose =3)

mergedColors = merge$colors
mergedMEs = merge$newMEs
```

```{r}
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors), c("Dynamic Hybrid Cut", "Merged Hybrid"), dendroLabels= FALSE, hang=0.03, addGuide= TRUE, guideHang=0.05)
```


```{r}
# set a threshold for merging modules
dynamicMergeCut(36) # calculates the threshold for module merging using the number of samples
```


```{r}
MEDissThres = 0.2131385
# Plot the cut line into the dendrogram
sizeGrWindow(12, 9)
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
abline(h=MEDissThres, col = "red")
```


```{r}
merge = mergeCloseModules(dat, dynamicColors, cutHeight = MEDissThres, verbose = 3)

# The merged module colors
mergedColors = merge$colors

# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs
```


```{r}
# Calculate dissimilarity of module eigengenes
MEDiss_merged = 1-cor(mergedMEs)

# Cluster module eigengenes
METree_merged = hclust(as.dist(MEDiss_merged), method = "average")

# Plot the result
sizeGrWindow(7, 6)
plot(METree_merged, main = "Clustering of MERGED module eigengenes",
     xlab = "", sub = "")
abline(h = MEDissThres, col="red")
```

## Clustering dendrogram of all genes, with dissimilarities based on topological overlap. 
Each vertical line represents a single gene. Assigned module colors below.
```{r}
sizeGrWindow(12, 9)
#pdf(file = "___.pdf", wi = 9, he = 6)
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
                    c("Dynamic Hybrid Cut", "Merged hybrid"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
```

save plot
```{r}
pdf(file="Sid_Host_WGCNA_MergeNetwork.pdf", width=20, height=20)
#pdf(file = "___.pdf", wi = 9, he = 6)

sizeGrWindow(12, 9)
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
                    c("Dynamic Hybrid Cut", "Merged hybrid"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)

dev.off()
```


```{r}
moduleColors_merged = mergedColors
colorOrder = c("grey", standardColors(50))
moduleLabels = match(moduleColors_merged, colorOrder)-1
#MEs_final = mergedMEs
save(mergedMEs, moduleLabels, moduleColors_merged, MEDiss_merged, METree_merged, geneTree, file= "Sid_Host_WGCNA_ME_Network_merged.RData")
```



## Quantify module similarity by eigengene correlation. 

Eigengenes: Module representatives
```{r}
plotEigengeneNetworks(mergedMEs, "", marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2))
```

## Look at expression patterns of these genes, as they're clustered
```{r}
module.order <- unlist(tapply(1:ncol(dat),as.factor(mergedColors),I))
m <- t(t(dat[,module.order])/apply(dat[,module.order],2,max))
heatmap(t(m),zlim=c(0,1),col=gray.colors(100),Rowv=NA,Colv=NA,labRow=NA,scale="none",RowSideColors=mergedColors[module.order])
```


####################################################################

# 3. Relate modules to external information

Array Information: clinical data, SNPs, proteomics 
Gene Information: ontology, functional enrichement
Rationale: find biologically interesting modules


# Relating modules to traits

Compute 1st principal componet of each module as its eigengenes.
Correlate eigengene external traits and look for the most significant associations.
```{r}
# Define numbers of genes and samples
nGenes = ncol(dat)
nSamples = nrow(dat)

# extract eigengenes (1st principal component) of modules
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(dat, moduleColors_merged, softPower = softPower)$eigengenes
MEs = orderMEs(MEs0)

# output is same
#MEs = orderMEs(mergedMEs)

# correlations of traits with eigengenes
moduleTraitCor = cor(MEs, traits, use = "p")
moduleTraitPvalue = corPvalueFisher(moduleTraitCor, nSamples)

# alternate
#moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)

# correlations of genes with eigengenes
moduleGeneCor=cor(MEs,dat)
moduleGenePvalue = corPvalueStudent(moduleGeneCor, nSamples)
```


## Display the correlation values within a heatmap plot
```{r}
# Open a graphic window with specified dimensions
sizeGrWindow(10,6)

# Will display correlations and their p-values
textMatrix = paste(signif(moduleTraitCor, 2), "\n(", signif(moduleTraitPvalue, 1), ")", sep = "")
dim(textMatrix) = dim(moduleTraitCor)

par(mar = c(6, 8.5, 3, 3))

labeledHeatmap(Matrix = moduleTraitCor, 
               xLabels = names(traits), 
               yLabels = names(MEs), 
               ySymbols = names(MEs), 
               colorLabels = FALSE, 
               colors = blueWhiteRed(50), 
               textMatrix = textMatrix, 
               setStdMargins = FALSE, 
               cex.text = 0.5, 
               zlim = c(-1,1), 
               main = paste("Module-trait relationships"))
```


## represent signifiant module trait correlations as a heatmap
```{r}
library(RColorBrewer)
modLabels=sub("ME","",names(MEs))

ps=signif(moduleTraitPvalue,1)
cors=signif(moduleTraitCor,2)
textMatrix = cors;
#paste(cors, "\n(",ps, ")", sep = "");
textMatrix[ps>0.05]="-"
dim(textMatrix) = dim(moduleTraitCor)

par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor,
xLabels = names(traits),
ySymbols = modLabels,
yLabels = modLabels,
colorLabels = FALSE,
colors = colorRampPalette(c("blue","lightblue","white","coral","red"))(50),
textMatrix = textMatrix,
setStdMargins = FALSE,
cex.text = 0.7,
zlim = c(-0.7,0.7),
main = paste("Siderastrea Host Module-Trait correlations"))
```


save plot
```{r}
pdf("Sid_Host_module-trait_heatmap.pdf")

par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor,
xLabels = names(traits),
ySymbols = modLabels,
yLabels = modLabels,
colorLabels = FALSE,
colors = colorRampPalette(c("blue","lightblue","white","coral","red"))(50),
textMatrix = textMatrix,
setStdMargins = FALSE,
cex.text = 0.7,
zlim = c(-0.7,0.7),
main = paste("Siderastrea Host Module-Trait correlations"))

dev.off()
```


```{r}
mct=table(moduleColors_merged)
mct[modLabels]
```

### module size barplot
```{r}
labelShift=300 # increase to move module size labels to the right
par(mar = c(6, 8.5, 3, 3))
x=barplot(mct[rev(modLabels)],horiz=T,las=1,xlim=c(0,3000),col=rev(modLabels))
text(mct[rev(modLabels)]+labelShift,y=x,mct[rev(modLabels)],cex=0.9)
```


save information about module genes
```{r}
#green <- as.data.frame(names(dat)[moduleColors_merged == "green"])

cyan <- names(dat)[moduleColors_merged == "cyan"]
cyan <- as.data.frame(cyan)
names(cyan)[names(cyan) == "cyan"] <- "gene"
cyan$module <- "cyan"

midnightblue <- names(dat)[moduleColors_merged == "midnightblue"]
midnightblue <- as.data.frame(midnightblue)
names(midnightblue)[names(midnightblue) == "midnightblue"] <- "gene"
midnightblue$module <- "midnightblue"

tan <- names(dat)[moduleColors_merged == "tan"]
tan <- as.data.frame(tan)
names(tan)[names(tan) == "tan"] <- "gene"
tan$module <- "tan"

green <- names(dat)[moduleColors_merged == "green"]
green <- as.data.frame(green)
names(green)[names(green) == "green"] <- "gene"
green$module <- "green"

brown <- names(dat)[moduleColors_merged == "brown"]
brown <- as.data.frame(brown)
names(brown)[names(brown) == "brown"] <- "gene"
brown$module <- "brown"

red <- names(dat)[moduleColors_merged == "red"]
red <- as.data.frame(red)
names(red)[names(red) == "red"] <- "gene"
red$module <- "red"

salmon <- names(dat)[moduleColors_merged == "salmon"]
salmon <- as.data.frame(salmon)
names(salmon)[names(salmon) == "salmon"] <- "gene"
salmon$module <- "salmon"

blue <- names(dat)[moduleColors_merged == "blue"]
blue <- as.data.frame(blue)
names(blue)[names(blue) == "blue"] <- "gene"
blue$module <- "blue"

lightcyan <- names(dat)[moduleColors_merged == "lightcyan"]
lightcyan <- as.data.frame(lightcyan)
names(lightcyan)[names(lightcyan) == "lightcyan"] <- "gene"
lightcyan$module <- "lightcyan"

greenyellow <- names(dat)[moduleColors_merged == "greenyellow"]
greenyellow <- as.data.frame(greenyellow)
names(greenyellow)[names(greenyellow) == "greenyellow"] <- "gene"
greenyellow$module <- "greenyellow"

magenta <- names(dat)[moduleColors_merged == "magenta"]
magenta <- as.data.frame(magenta)
names(magenta)[names(magenta) == "magenta"] <- "gene"
magenta$module <- "magenta"

yellow <- names(dat)[moduleColors_merged == "yellow"]
yellow <- as.data.frame(yellow)
names(yellow)[names(yellow) == "yellow"] <- "gene"
yellow$module <- "yellow"

black <- names(dat)[moduleColors_merged == "black"]
black <- as.data.frame(black)
names(black)[names(black) == "black"] <- "gene"
black$module <- "black"

purple <- names(dat)[moduleColors_merged == "purple"]
purple <- as.data.frame(purple)
names(purple)[names(purple) == "purple"] <- "gene"
purple$module <- "purple"

summary <- rbind(cyan, midnightblue, tan, green, brown, red, salmon, blue, lightcyan, greenyellow, magenta, yellow, black, purple)

write.csv(file = "Sid_Host_WGCNA_module_genes.csv", x = summary)
```



####################################################################


# 4. Study module relationships
Rationale: biological data reduction, systems-level view 
Tools: Eigengene Networks


Interfacing network analysis with other data such as functional annotation and gene ontology
https://systemsbio.ucsd.edu/WGCNAdemo/


####################################################################

# 5. Find the key drivers in interesting modules 
Rationale: experimental validation, biomarkers 
Tools: intramodular connectivity, causality testing



##############################################################################



```{r}
sessionInfo()
```


